File: service/src/users/usecases/lookup_username.rs
use crate::users::{Username, UsersService};

impl UsersService {
    /// Look up a Username to see if it's known to the system
    ///
    /// # Parameters
    /// - `username` - The username to look up
    ///
    /// # Returns
    /// True if the username exists. False if not
    pub fn lookup_username(&self, username: &Username) -> bool {
        username == &Username::new("known")
    }
}

File: webapp/src/api/authentication/lookupUsername.ts
import { Problem, request } from "../http";

import debug from "debug";

/** The logger to use */
const LOGGER = debug("multiverse:landing:authentication:api");

/**
 * Actually look up the username to see if it exists
 * @param username The username to look up
 */
export async function lookupUsername(username: string): Promise<boolean> {
  try {
    const response = await request("/usernames/{username}", {
      urlParams: {
        username,
      },
    });
    LOGGER("Username exists: %o", response);
    return true;
  } catch (e) {
    if (
      e instanceof Problem &&
      e.type === "tag:multiverse,2020:users/problems/unknown_username"
    ) {
      LOGGER("Username doesn't exist");
      return false;
    } else {
      LOGGER("Something went wrong: %o", e);
      throw e;
    }
  }
}
